MAINT, WIP: move extra attributes to _raw_extras in eyelink reader

scott-huberty opened the pull request:
Fixes issue that was raised in #11825 and specifically [here](https://github.com/mne-tools/mne-python/pull/11823#discussion_r1271291973).

@larsoner there are quite a few extra attributes in `read_raw_eyelink` to move to `_raw_extras[0]`. In this initial commit I've only made the change for a single attribute, to keep things simple. Hoping you can review it when you have time and confirm that this is what we want. Once you do I'll move the rest of the unwanted attributes to `_raw_extras[0]`

Thanks!

larsoner commented on mne/io/eyelink/eyelink.py:
```diff
@@ -210,8 +210,8 @@ def __init__(
         self.fname = Path(fname)
         self._sample_lines = None  # sample lines from file
         self._event_lines = None  # event messages from file
+        self._raw_extras = list([dict()])  # extra info from file
```
```suggestion
        raw_extras = dict()  # extra info from file
```

scott-huberty replied:
Thanks for the tips! I can make raw_extras a variable as you suggest here, but from what I can tell I'll just need to define it outside of the `RawEyelink` class so that it is accessible to the various methods/funcs that access it before the `Super` call.

larsoner commented on mne/io/eyelink/eyelink.py:
```diff
@@ -267,9 +267,13 @@ def __init__(
+            preload=eye_ch_data,
+            filenames=[self.fname],
+            verbose=verbose,
+            raw_extras=self._raw_extras,
```
```suggestion
            raw_extras=[raw_extras],
```
The super method will set the `self._raw_extras` attribute. See for example how other readers do it like 

https://github.com/mne-tools/mne-python/blob/1ee327014b9389962b70331f41b7de979e01140b/mne/io/brainvision/brainvision.py#L111-L119

larsoner commented on mne/io/eyelink/eyelink.py:
```diff
@@ -398,7 +402,7 @@ def _get_recording_datetime(self):
                         dt_aware = dt_naive.replace(tzinfo=tz)
-                        self._meas_date = dt_aware
+                        self._raw_extras[0]["dt"] = dt_aware
                         break
```
And ideally this would become a private function like `def _get_recording_datetime(fname): ... retun dt` rather than a method

scott-huberty commented:
In order to make `mne.io.tests.test_raw._test_raw_reader` happy (which is basically the scope of this PR), and to follow your advice [here](https://github.com/mne-tools/mne-python/pull/11910#discussion_r1303011336), I had to do restructuring. 

Before, I was using a class `RawEyelink` to parse the incoming file, create the annotations, and pass everything to `BaseRaw`, making use of a lot of custom private methods. Now I've converted all those private methods to helper funcs and moved them to a utils folder. I tried my best not to make anything more complex than it was before. 

Now at least passing `RawEyelink` to `mne.io.tests.assert_attributes` would pass (there aren't any extra methods or attributes in `RawEyelink`). There are still some other failures in `mne.io.eyelink.tests.test_eyelink.test_basics` that I'll try to figure out locally before removing the `pytest.marks.xfail`  from that test.

larsoner commented:
Sounds good -- that's much more like how other `io/` readers behave so I think it's helpful!

scott-huberty commented:
Right now, [mne.io.eyelink.tests.test_eyelink.test_basics](https://github.com/mne-tools/mne-python/blob/e94f255675c9287e8f69ca08db20a1eb6ada8a69/mne/io/eyelink/tests/test_eyelink.py#L287) fails [here](https://github.com/mne-tools/mne-python/blob/e94f255675c9287e8f69ca08db20a1eb6ada8a69/mne/io/tests/test_raw.py#L223-L224)

```
with pytest.raises(RuntimeError, match="loaded"):
    apply.copy().pick(picks[0]).get_data()
```

Because It seems like `RawEyelink` instances are always preloaded..

```
import mne
from mne.datasets.eyelink import data_path
from mne.utils import _check_preload

et_fpath = data_path() / "sub-01_task-plr_eyetrack.asc"
eeg_fpath = data_path() / "sub-01_task-plr_eeg.mff"

raw_et = mne.io.read_raw_eyelink(et_fpath, preload=False)
raw_eeg = mne.io.read_raw_egi(eeg_fpath, preload=False)

_check_preload(raw_et, "cheking preload")  # Does not raise
_check_preload(raw_eeg, "cheking preload") # Raises
```

I still am trying to figure out why this is the case.

larsoner commented:
If data are always preloaded you should pass `test_preloading=False` to `_test_raw_reader`, and `read_raw_eyelink` should not have a `preload=False` option actually...

larsoner commented:
> Got it! It's okay for RawEyelink to always be preloaded?

Ideally there would be an option not to preload. But if it's unlikely to be used by anyone (which seems to be the case based on file sizes) and complicated to add (which will probably be the case, at least in a way that actually saves memory) then we don't need to. We can always add it later...

But in the meantime we should get rid of the `preload` argument. We can just get rid of it entirely since it's a bugfix

scott-huberty commented:
> ` [03da908](/mne-tools/mne-python/pull/11910/commits/03da908bde4a2975421291efa137473c6481b845)`

That's about the extent of it so that's fine!

larsoner commented:
Thanks @scott-huberty !

scott-huberty commented on mne/io/tests/test_raw.py:
```diff
@@ -319,8 +319,8 @@ def _test_raw_reader(
     assert raw.__class__.__name__ in repr(raw)  # to test repr
     assert raw.info.__class__.__name__ in repr(raw.info)
     assert isinstance(raw.info["dig"], (type(None), list))
-    data_max = full_data.max()
-    data_min = full_data.min()
```
@larsoner I did make this change in `test_raw` that I just want to make sure you are aware of.

larsoner replied:
Yes I saw this one and I think it's fine :)
